ruby/extractor/src/diagnostics.rs

```diff
@@ -0,0 +1,255 @@
+    pub telemetry: bool,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Location {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /** Path to the affected file if appropriate, relative to the source root */
+    pub file: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub start_line: Option<i32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub start_column: Option<i32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub end_line: Option<i32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub end_column: Option<i32>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DiagnosticMessage {
+    /** Unix timestamp */
+    pub timestamp: u64,
+    pub source: Source,
+    #[serde(skip_serializing_if = "String::is_empty")]
+    /** GitHub flavored Markdown formatted message. Should include inline links to any help pages. */
+    pub markdown_message: String,
+    #[serde(skip_serializing_if = "String::is_empty")]
+    /** Plain text message. Used by components where the string processing needed to support Markdown is cumbersome. */
+    pub plaintext_message: String,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    /** List of help links intended to supplement the `plaintextMessage`. */
+    pub help_links: Vec<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub severity: Option<Severity>,
+    #[serde(skip_serializing_if = "std::ops::Not::not")]
+    pub internal: bool,
+    #[serde(skip_serializing_if = "is_default_visibility")]
+    pub visibility: Visibility,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub location: Option<Location>,
+}
+
+fn is_default_visibility(v: &Visibility) -> bool {
+    !v.cli_summary_table && !v.status_page && !v.telemetry
+}
+
+pub struct LogWriter {
+    path: Option<PathBuf>,
+    inner: Option<std::io::BufWriter<std::fs::File>>,
+}
+
+impl LogWriter {
+    pub fn write(&mut self, mesg: &DiagnosticMessage) -> std::io::Result<()> {
+        if self.inner.is_none() {
+            let mut open_failed = false;
+            self.inner = self.path.as_ref().and_then(|path| {
+                match std::fs::OpenOptions::new()
+                    .create(true)
+                    .append(true)
+                    .write(true)
+                    .open(&path)
+                {
+                    Err(e) => {
+                        tracing::error!(
+                            "Could not open log file '{}': {}",
+                            &path.to_string_lossy(),
+                            e
+                        );
+                        open_failed = true;
+                        None
+                    }
+                    Ok(file) => Some(std::io::BufWriter::new(file)),
+                }
+            });
+            if open_failed {
+                self.path = None
```

Comment:
That what I tried before but I got the following error:
```
error[E0500]: closure requires unique access to `self` but it is already borrowed
   --> extractor/src/diagnostics.rs:125:54
    |
125 |             self.inner = self.path.as_ref().and_then(|path| {
    |                          ---------          -------- ^^^^^^ closure construction occurs here
    |                          |                  |
    |                          |                  first borrow later used by call
    |                          borrow occurs here
...
138 |                         self.path = None;
    |                         --------- second borrow occurs due to use of `self` in closure
```
There may be a better way to avoid that though.

Reply:
That looks great, thanks!